### ISSUE TITLE
Warning about low entropy isn't completely correct

### ISSUE THREAD

**Issue by theblazehen** [firehol/netdata#2510]
> If the pool of entropy gets empty, you risk your security to be predictable and you should install a user-space random numbers generating daemon, like haveged or rng-tools (i.e. rngd), to keep the pool in healthy levels.

See https://www.2uo.de/myths-about-urandom/

> 
> What about entropy running low?
> 
> It doesn't matter.
> 
> The underlying cryptographic building blocks are designed such that an attacker cannot predict the outcome, as long as there was enough randomness (a.k.a. entropy) in the beginning. A usual lower limit for “enough” may be 256 bits. No more.
> 
> Considering that we were pretty hand-wavey about the term “entropy” in the first place, it feels right. As we saw, the kernel's random number generator cannot even precisely know the amount of entropy entering the system. Only an estimate. And whether the model that's the basis for the estimate is good enough is pretty unclear, too. 

I'd suggest replacing "you risk your security to be predictable" with something about how it can cause applications to block waiting for entropy from /dev/random

**Comment by Ferroin** [comment]
If you need random numbers for mathematical applications (not cryptographic), entropy is important though.

That said, most people don't fall into that category, and I would at least advocate changing the message on two other grounds:
1. HAVEGE doesn't work reliably on many small non-x86 systems, because it partially relies on timer skew that can only be present if there is more than one oscillator in the system.  It is also _not_ cryptographically secure, and predictable input to a RNG results in predictable output.
2. rngd is pointless on recent Linux versions.  The kernel already feeds data from hardware RNG's into the entropy pool itself, so rngd can actually make things worse (putting multiple copies of the same data into the pool technically _reduces_ the overall entropy by a small percentage), not to mention that it wastes CPU time running FIPS tests on the data from the hardware RNG.

**Comment by theblazehen** [comment]
@Ferroin How would that matter, since a CSPRNG needs to be indistinguishable from true randomness for it to be useful. Also, /dev/random doesn't give you direct randomness either, 
>    Fact: Both /dev/urandom and /dev/random are using the exact same CSPRNG (a cryptographically secure pseudorandom number generator). They only differ in very few ways that have nothing to do with “true” randomness.

------------------------

>  It is also not cryptographically secure, and predictable input to a RNG results in predictable output.
> (putting multiple copies of the same data into the pool technically reduces the overall entropy by a small percentage),

That shouldn't matter though, unless you have an actively malicious source of entropy, 

> It has been suggested that Intel’s hardware number generator may not be fully random. Since it is baked into the silicon, that assertion is hard to audit and verify. As it turns out, even if the numbers generated have some bias, it can still help as long as this is not the only source of randomness in the system. Even if the random number generator itself had a back door, the mixing property of randomness means that it cannot lower the amount of entropy in the pool.

https://blog.cloudflare.com/ensuring-randomness-with-linuxs-random-number-generator/ and https://blog.cr.yp.to/20140205-entropy.html

**Comment by Ferroin** [comment]
A PRNG is just that, a _pseudo_ random number generator.  CSPRNGS are deterministic without sufficient input entropy (especially if they are compliant with any kind of government standard on randomness), and aren't truly random because they actively prevent things like runs of the same value that can and should occur for a true random number generator (but are bad for cryptographic purposes).

On the note of HAVEGE's cryptographic security, it relies on non-determinism in the internal CPU state as an entropy source.  That non-determinism is entirely based on userspace context on the system, which means it's possible to manipulate it without much effort, especially if you have knowledge of the CPU internals (or if they are very simple).

The issue with duplicate data (which you've quoted in a different context than it originated I might add) arises from things being time correlated _and_ duplicated.  This is the same as the argument on LKML at the moment about how the kernel handles HID events and disk interrupts WRT the entropy pool (both are effectively counted twice, because they both get counted by the regular interrupt entropy harvesting code).

**Comment by ktsaou** [comment]
So,

netdata today says:

> Entropy, is like a pool of random numbers (/dev/random) that are mainly used in cryptography. It is advised that the pool remains always above 200. If the pool of entropy gets empty, you risk your security to be predictable and you should install a user-space random numbers generating daemon, like haveged or rng-tools (i.e. rngd), to keep the pool in healthy levels.

I can change it like this:

> Entropy, is like a pool of random numbers (/dev/random) that are mainly used in cryptography. It is advised that the pool remains always above 200. If the pool of entropy gets empty, processes requiring random numbers will run a lot slower, waiting for the pool to be replenished.

It would be great though to give some advice on what someone can do to fix the issue (I have several systems that frequently get out of entropy, and normally I install either haveged or rngd to fix it). What is the recommended way?

**Comment by theblazehen** [comment]
@Ferroin 

>  and aren't truly random because they actively prevent things like runs of the same value that can and should occur for a true random number generator (but are bad for cryptographic purposes).

Source? Also, /dev/random comes from the same source as /dev/urandom, as I quoted, so if /dev/urandom doesn't work for you then neither will /dev/random. And if the kernel was removing repeating bytes from /dev/urandom then it would be reducing randomness. If a hash function returns some repeating data after hashing some data, then that's just chance, you don't then throw the hash out because you don't think it looks random enough. 

RE HAVEGE your initial statement made it sound like adding data from a CSPRNG to the entropy pool would be bad, which it wouldn't, however only eg loading a kernel module or having a compromised hardware randomness generator with access to all memory could have an impact with malicious randomness, see https://github.com/torvalds/linux/blob/master/drivers/char/random.c#L81-L99

And again, since all randomness is hashed together, only an attacker that can see the internal state of the RNG (hardware backdoor / kernel module) the duplicate data won't make it worse.

--

@ktsaou That sounds good. The recommended way would be for you to get the application to use /dev/urandom instead of /dev/random

For fixing it, without changing anything in the application, see https://lwn.net/Articles/525459/

**Comment by Ferroin** [comment]
First off, I wouldn't mention /dev/random specifically.  getrandom(2) also uses the kernel's entropy pool, as do some other kernel-internal functions.  Second, once it's seeded, it's not horrible to run low as long as it's not constantly running low while the system is regularly generating random numbers (so it's not too bad on a client node, but it can be pretty bad on a very busy web server that sees lots of TLS connections).

On the note of dealing with it, there are multiple options, but the biggest one is to figure out what's eating all your entropy, and deal with that.  No properly written software should be using /dev/urandom or /dev/random for anything beyond seeding it's own RNG.  If you running software that's legitimately using it and running you out of entropy, you need to be looking at getting a system with a hardware RNG, not installing some other software to try and handle this for you.  Even just getting a system with a TPM is sufficient (provided you have the TPM enabled (it doesn't need to be activated, and you don't need Secure Boot enabled or anything, you just need the TPM enabled so the OS can talk to it's RNG block)).

As far as software to install, haveged is usually fine on x86 systems, as well as bigger PPC or SPARC systems, but I would not advocate running it on small ARM or MIPS systems.  Rngd is outdated on modern kernels (they feed in entropy themselves), and doesn't help at all if you don't have a hardware RNG on the system.  There are a couple of other options too (timer_entropyd is decent on bigger systems, but you need a system with at least two oscillators for it to work, there are also a couple of decent options that use sound hardware to collect environmental entropy, but they require a sound card with an analog input (and ideally a microphone)).  The problem is though that any of the options that do exist are only good choices in some circumstances, and making a bad choice there can make things _worse_ than having low entropy (adding bad entropy on a system that is already starved for entropy is never a good thing).

**Comment by ktsaou** [comment]
ok,

How about this:

> Entropy, is a pool of random numbers that is mainly used in cryptography. It is advised that the pool remains always above 200. If the pool of entropy gets empty, processes requiring random numbers will run a lot slower, waiting for the pool to be replenished. Ideally a system with high entropy demands should have a hardware device for that purpose (TPM is one such device). There are also several software-only options you may install, like `haveged`, although these are generally useful in servers (not IoT).

**Comment by Ferroin** [comment]
I would say that covers things pretty well, although 200 its seems kind of arbitrary.  Once it's seeded (which IIRC needs 256 bits on Linux, not sure on *BSD), it's not all that crucial to keep it high unless you use it a lot.  Even then, most of the systems this is likely to be low on (client nodes that sit mostly idle and tiny IoT devices) don't often need it to be high frequently simply because they aren't doing _that_ much cryptography.

In all honesty, tracking the entropy pool size is cool and all, but it's really not a very useful statistic except for finding suspect crypto software or correlating with other data to determine usage patterns.

**Comment by ktsaou** [comment]
So you are suggesting to also remove the alarm?
I think we should have an alarm in place. Somehow people should know that their system might be slow because entropy is low...

**Comment by theblazehen** [comment]
>  No properly written software should be using /dev/urandom or /dev/random for anything beyond seeding it's own RNG.  If you running software that's legitimately using it and running you out of entropy,

But you don't need to consume a byte of entropy for each byte of random data you get out, that's the whole idea of a CSPRNG. So if you read from /dev/urandom then you won't run low. 

>  that. No properly written software should be using /dev/urandom or /dev/random for anything beyond seeding it's own RNG.
Why not? There's a reason it's a CSPRNG, and it's good enough for https://github.com/python/cpython/blob/9f396b605ec03fc5af100222f1b0a9063ac02a0b/Lib/random.py#L44 https://github.com/golang/go/blob/964639cc338db650ccadeafb7424bc8ebb2c0f6c/src/crypto/rand/rand.go#L14 and this bit from rust has a nice comment: https://github.com/rust-lang/rust/blob/ff9f2d2ae94fd951229c33ae55076fce4f68fab0/src/libstd/rand/mod.rs#L30-L57

Although I agree that having more / better randomness is always better, if someone has the ability to load a kernel module (to observe the state of the CSPRNG) or can implement backdoors in your hardware RNG then you have bigger problems. Once the CSPRNG has been seeded by some real random data (so not early boot if your distro doesn't save random data to disk that gets loaded in early boot) then the malicious kernel module / hardware backdoor becomes much more likely a form of attack than the RNG. which if it already has entropy available, won't be harmed by anything unless it's actively harmful.

Basically, you shouldn't be removing entropy when you're requesting random data, you should just read from the CSPRNG and let the kernel mix in new entropy as it becomes available

**Comment by theblazehen** [comment]
I think having the alarm is good just to let people know that programs may be blocking on getting randomness, as any program using getrandom() or /dev/random will block when entropy runs out

**Comment by Ferroin** [comment]
If the software is reading from /dev/urandom, then the entropy value is (debatably mostly) irrelevant.  If it's reading from /dev/random, then the software itself is why the system is blocking, not the lack of entropy.  If it's using getrandom, then unless it's going out of it's way to only work if the entropy pool has data (which is still bad practice), then it will only block early in boot before the pool is seeded)   IOW, either it's the software's fault for poor design, or it's too early for netdata to be realistically running, and neither case has anything to do with entropy on the system.

This brings to mind one of the few things I really admire about OpenBSD.  There, they have one device (/dev/arandom), and /dev/random and /dev/urandom are symlinks to that.  It provides a moderately high-bandwidth stream of cryptographically secure random numbers (it uses a custom ARC4-based CSPRNG based on a similar idea to the AES-OFB based gold-standard for CSPRNG's), doesn't block, and doesn't have all this entropy tracking crap that Linux has.

**Comment by ktsaou** [comment]
I see that after all this interesting discussion, we haven't got a decision to change anything in netdata.

So, I plan to change the message to:

> Entropy, is a pool of random numbers that is mainly used in cryptography. If the pool of entropy gets empty, processes requiring random numbers may run a lot slower (it depends on the interface the program uses), waiting for the pool to be replenished. Ideally a system with high entropy demands should have a hardware device for that purpose (TPM is one such device). There are also several software-only options you may install, like haveged, although these are generally useful in servers.

And I will not remove or change the alarm.

Do we agree?

**Comment by ktsaou** [comment]
here it is.

**Comment by theblazehen** [comment]
Agreed on that
